pkg/logservice/client.go

```diff
@@ -157,18 +165,33 @@ func (c *client) request(ctx context.Context,
 		MaxSize:     maxSize,
 		PayloadSize: uint64(len(payload)),
 	}
-	if err := writeRequest(c.conn, req, c.buf, payload); err != nil {
+
+	// TODO: check with ZhangXu whether this is safe
+	rr := c.req
+	rr.Request = req
+	rr.payload = payload
+	future, err := c.client.Send(ctx, c.addr, rr,
+		morpc.SendOptions{Timeout: time.Duration(timeout)})
+	if err != nil {
 		return pb.Response{}, nil, err
 	}
-	resp, recs, err := readResponse(c.conn, c.buf)
+	msg, err := future.Get()
 	if err != nil {
 		return pb.Response{}, nil, err
 	}
-	err = toError(resp)
+	response, ok := msg.(*RPCResponse)
+	if !ok {
+		panic("unexpected response type")
+	}
+	var recs pb.LogRecordResponse
+	if response.PayloadSize > 0 {
```

Comment:
PayloadSize field not needed anymore. The Payload field can be defined in the RPCResponse, and the PayloadMessage interface can be implemented to support the transmission of Payload